### DISCUSSION TITLE
LazyLoading + pass params to view model of TabViewItems MVVM first approach.

### DISCUSSION THREAD

**Discussion by LeoJHarris** [xamarin/XamarinCommunityToolkit#1529] [Q&A] [answered]
Hi all, I am trying to implement LazyView on my TabViewItem. My Xamarin project is using Prism and I have a main page that contains my TabView. I would like my TabViewItems to utilize LazyView. 

My main page is passed a parameter which needs to be presented to the ViewModels of the tab items as there are some methods that get called when loaded, how can I pass and provide the paramater to the TabViewItems in the VM that are wrapped with LazyView and do this once View is Loaded? @pictos 

Something like the prism equivalent is what I need, I cannot use regions in Templated View and doesn't provide lazy loading hence I am keen to get this working:

```csharp
            if (parameters.ContainsKey("customerid"))
            {
                _regionManager.RequestNavigate("RefillsRegion", "RefillsView", RegionNavigationCallback, new NavigationParameters
                        {
                            { "customerid", parameters.GetValue<long>("customerid") }
                        });
            }
        }
```

Thanks in advance

**Comment by pictos** [comment]
@LeoJHarris I moved this to the Discussion session since this fits better. So I don't have too much experience with prism and other MVVM frameworks. Are you able to provide a simple sample project that I can play with and see if I find a good solution?

I believe that you will need to create your own LazyView implementation with those extra features

**Reply by LeoJHarris** [reply]
Yes I can create a small simple sample, will get that shortly.

**Reply by LeoJHarris** [reply]
Hi @pictos thanks for taking the time to check this out for me.

Attached is a very simple case where I want to take the parameter passed to page `TabsPageViewModel` and then provide it to the `NestedTabViewModel` which is the view model behind the view that is wrapped with LazyView on the TabViewItem. 

Some things to note I commented out a Constructor that I would like to use for constructor injection with the iOC container if possible

`value` is what I want in my `NestedTabViewModel` when loaded as its a parameter that might be used when page is reloaded and not simply Initialized either.

![image](https://user-images.githubusercontent.com/5468757/126582609-7e507727-cc77-4407-a4fe-4b9b0005eb3c.png)

![image](https://user-images.githubusercontent.com/5468757/126582393-fdd7ced8-4705-4b1d-891d-42d5c93c260b.png)

[Sample.zip](https://github.com/xamarin/XamarinCommunityToolkit/files/6859631/Sample.zip)

**Reply by pictos** [reply]
@LeoJHarris great! I'll take a look on this until Friday (:

**Reply by LeoJHarris** [reply]
@pictos thanks although its using prism its not really adding much more complexity the problem hopefully

**Reply by pictos** [reply]
![image](https://user-images.githubusercontent.com/20712372/126585816-346a7a09-33a5-46e0-b276-79a3ab10f632.png)

Have been a fun distraction, I already make the ViewModel ctor work, now is just passing the navigation value into `NestedTabViewModel`

**Reply by LeoJHarris** [reply]
@pictos Sounds good it was a bit of problem I couldnt quite get a solution on. I wasn't sure how to make the ViewModel ctor work as in the example code you were initializing a new instance to `Current` and I am pretty sure that is to be avoided as it cannot be unit tested and mocked. Should be initialized through dependency injection with iOC container.

For example here:

https://github.com/xamarin/XamarinCommunityToolkit/blob/365b0cf316afded845ca094b1abe1d8357d9ab64/samples/XCT.Sample/ViewModels/Views/Tabs/LazyTestViewModel.cs#L7

**Reply by pictos** [reply]
![image](https://user-images.githubusercontent.com/20712372/126587080-6291551f-8fca-4224-afbd-6f8c40457401.png)
And we have a winner, I'll make the code more pretty and I'll upload the zip file.

Yeah... This is a very silly code, sorry (but not sorry) for that, that is just to give you an idea of how to do it. Sadly, to set the `BindingContext` (BC) for LazyView you need a `static instance`. If you want to do it in another way you will need to implement your own `LazyView` that supports it. The history behind it is that I want to provide a very simple `LazyView` for people, is hard-sh to guess how people will want to use it.. Maybe in the future we can have more LazyView flavors or just make this more robust, let's see...

**Reply by LeoJHarris** [reply]
@pictos thank you for this! I can understand all of that. keen to see the implementation 👍  thank you @pictos 🔥 and added you to my following list 😄

**Reply by pictos** [reply]
It's bedtime for me now. If you have any other issues with the sample just post here and I'll take a look tomorrow. Have a great day \o/

**Reply by LeoJHarris** [reply]
Thank you @pictos

**Reply by LeoJHarris** [reply]
@pictos thanks again for the time taken to help me on this, that solution works good for me. This is part of my actual application now and just wondering to keep this as lazy loading am I able to execute a command on my tab view model that will take a value and do some tasks required within the view model i.e. calling some API's or do I just set a property on the view model and then wait for `isLoaded` to be called and  execute some code with the value?

Something like the below where once the page is initialized with the parameter I can call RefreshTabViewModelContentAsync on the view model for which the tab view is being lazy loaded.

```csharp
public override void Initialize(INavigationParameters parameters)
        {
            base.Initialize(parameters);

            if (parameters.ContainsKey("customerid"))
            {
                RefillsTabViewModel.RefreshRefillsCommand.Execute(customerid);
                HistoryTabViewModel.RefreshScriptHistoryCommand.Execute(customerid);
            }
        }
```

Should the refresh commands be getting called when `isLoaded` is true?

**Reply by pictos** [reply]
@LeoJHarris Sorry for the late reply... What's the behavior that you want in the UI? If I remember well, the VM will be created before the View (LazyView), so when the View is rendered the RefreshCommands are already triggered. So this behavior is causing any issue?

**Reply by LeoJHarris** [reply]
@pictos sorry for the really late response on this I think its all sorted.

**Reply by pictos** [reply]
@LeoJHarris it's fine. Can I close this discussion?

**Reply by LeoJHarris** [reply]
@pictos yup conversation resolved 😄

**Comment by pictos** [comment, marked as answer]
[Sample.zip](https://github.com/xamarin/XamarinCommunityToolkit/files/6859883/Sample.zip)


@LeoJHarris

**Comment by HASSANRAGAI** [comment]
@LeoJHarris 
I don't know if you still want to solve it   I had the same issue and I edited the @pictos solution and you don't have to set the ViewModel manual 
see this repo and feel free to help me if there are any edits

https://github.com/ZHEAISNSAABN/ZhEaIsNsAaBn.Xamarin/tree/master/Src/ZhEaIsNsAaBn.Xamarin/LazyView

**Reply by LeoJHarris** [reply]
@HASSANRAGAI great work I had a brief look at the setup and looks great.

**Reply by HASSANRAGAI** [reply]
@LeoJHarris thnx hope it will help other developers
